File: 0146-lru-cache/0146-lru-cache.py
'''
capacity is always positive and also might be 0 ?
How big the capacity can be ?
hash_map get fetches in o(1)
so lets say we get the key value we will swap that with the first 

cap = 2    
ele = [1, 2]
get(1)
ele = [1 , 2]
get(2)
ele = [2, 1]
put(3)
the least will be 1 and when i pop that i will take o(1) 
and when i insert i insert at the 0th position so put takes o(1)

'''


# class OrderedDict:
#     def __init__(self):
#         self.store = {}
#         self.order = []
    
#     def move_to_end(self, key):
#         if key not in self.order:
#             return -1
#         else:
#             self.order.remove(key)
#             self.order.append(key)
            
#     def popitem(self, last):
#         if last:
#             key = self.order.pop()
#         else:
#             key = self.order.pop(0)
#         self.store.pop(key)
    
#     def __len__ (self):
#         return len(self.store)
    
#     def __contains__ (self, key):
#         return key in self.store

#     def __getitem__(self, key):
#         if key in self.store:
#             return self.store[key]
#         else:
#             return -1
    
#     def __setitem__(self, key, value):
#         if key not in self.store:
#             self.order.append(key)
#         self.store[key] = value
class ListNode:
    def __init__(self, key, val):
        self.next = None
        self.prev = None
        self.key = key
        self.val = val

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.head = ListNode(-1, -1)
        self.tail = ListNode(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.dic = {}
       
    def get(self, key: int) -> int:
        if key in self.dic:
            node = self.dic[key]
            del self.dic[key]
            self.remove(node)
            self.add(node.key, node.val)
            return node.val
        else:
            return -1
            
    def put(self, key: int, value: int) -> None:
        if key in self.dic:
            self.remove(self.dic[key])
        self.add(key, value)
        if  len(self.dic) > self.capacity:
            # we should remove the least recent use
            node = self.head.next
            self.remove(node)
            del self.dic[node.key]
           
    
    def add(self, key, value):
        node = ListNode(key, value)
        self.dic[key] = node
        prev_node = self.tail.prev
        prev_node.next = node
        node.prev = prev_node
        node.next = self.tail
        self.tail.prev = node

    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.java
/*
Capacity : 2

{2 : Node2, 3 : Node3 }

doubly linked list
h - > Node - >  Node -> tail

h - >Node1-> Node3 -> tail
Node2 = prev key val next
O(1)
O(1)
*/ 
class Node{
    int key, value;
    Node prev, next;
    Node(int key, int value){
        this.key = key;
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}
class LRUCache {
    private HashMap<Integer, Node> map;
    private Node head, tail;
    private int capacity;
    public LRUCache(int capacity) {
        this.map = new HashMap<>(capacity);
        this.head = new Node(-1, -1);
        this.tail = new Node(-1, -1);
        this.head.next = tail;
        this.tail.prev = head;
        this.capacity = capacity;
    }
    
    public int get(int key) {
        if(!map.containsKey(key)){
            return -1; 
        }
        else{
            Node node = map.get(key);
            remove(node);
            add(node);
            return node.value; 
        }
        
    }
    
    public void put(int key, int value) {
        if(!map.containsKey(key)){
            Node newNode = new Node(key, value);
            if (map.size() == capacity){
                remove(head.next);
            }
            add(newNode);
        }
        else{
            Node presentNode = map.get(key);
            presentNode.value = value;
            remove(presentNode);
            add(presentNode);
        }
    }

    public void remove(Node node)
    {
        Node prevNode = node.prev;
        prevNode.next = node.next;
        node.next.prev = prevNode;
        map.remove(node.key);
    }
    public void add(Node newNode){
        Node prevNode = tail.prev;
        prevNode.next = newNode;
        newNode.prev = prevNode;
        tail.prev = newNode;
        newNode.next = tail;
        map.put(newNode.key, newNode);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */